### DISCUSSION TITLE
Working compilers

### DISCUSSION THREAD

**Discussion by FactoryXCode** [rmesch/Bitmaps2Video-for-Media-Foundation#1] [General]
Hi Renate,

The code doesn't work on for example Delphi XE7 and maybe above.
For example: 

```
procedure TDirectoryTree.GetAllFiles in unit uDirectoryTree

this code will make it work for D XE7:

procedure TDirectoryTree.GetAllFiles(const aStringList: TStringlist;
const
  aFileMask: string);
var
  FilePath,
  mask,
  SearchStr: string;
  MaskLen,
  MaskPos,
  SepPos: integer;

{$IF COMPILERVERSION < 30.0}
  i: Integer;
  Strings: TArray<string>;
  ClassicStrings: TStringDynArray;
{$ENDIF}

begin
  FilePath := IncludeTrailingBackSlash(GetFullFolderName(Selected));
  Assert(Assigned(aStringList));
  aStringList.Clear;
  mask := aFileMask;
  MaskLen := Length(mask);
  MaskPos := 0;

  while MaskPos >= 0 do
  begin
    SepPos := Pos(';', mask, MaskPos + 1) - 1;
    if SepPos >= 0 then
      SearchStr := Copy(mask, MaskPos + 1, SepPos - MaskPos)
    else
      SearchStr := Copy(mask, MaskPos + 1, MaskLen);


{$IF COMPILERVERSION > 29.0}
    aStringList.AddStrings(TDirectory.GetFiles(FilePath,
                                               SearchStr,
                                               TSearchOption.soTopDirectoryOnly));
{$ELSE}
    ClassicStrings := TDirectory.GetFiles(FilePath,
                                          SearchStr,
                                          TSearchOption.soTopDirectoryOnly);
    // Copy all fields to the new array
    for i := Low(ClassicStrings) to High(ClassicStrings) do
      Strings[i] := ClassicStrings[i];

    aStringList.AddStrings(Strings);
{$ENDIF}

    if (SepPos >= 0) then
      begin
        inc(SepPos);
        if (SepPos >= MaskLen) then
          SepPos := -1;
      end;
    MaskPos := SepPos;
  end;

  // Natural sorting order, e.g. '7' '8' '9' '10'
  aStringList.CustomSort(LogicalCompare);
end;
```

And some more issues like the declaration of this method:

```
function GetResamplingTask(const RTS: TResamplingThreadSetup; Index: integer;
  AlphaCombineMode: TAlphaCombineMode): TProc;
begin
  Result := procedure
    var
      y, ymin, ymax: integer;
      CacheStart: PBGRAInt;
    begin
      CacheStart := @RTS.CacheMatrix[Index][0];
      ymin:= RTS.ymin[Index];
      ymax:= RTS.ymax[Index];
      for y := ymin to ymax do
      begin
        ProcessRow(y, CacheStart, RTS, AlphaCombineMode);

      end; // for y
    end; // procedure
end;
```

This will produce the error: [dcc32 Error] uScaleWMF.pas(144): E2555 Cannot capture symbol 'RTS'

To solve this error, the function parameters should be like this:

```
function GetResamplingTask(RTS: TResamplingThreadSetup; 
                                              Index: integer;
                                              AlphaCombineMode: TAlphaCombineMode): TProc;
```

Note the first parameter that is declared as const. This would be ok using guids, pointers or other types that shouldn't be affected while processing, but not for records  containing methods or interfaces that would affect the reference counting mechanism . 

An other issue is the declaration of IntToHex(hr), this should be IntToHex(hr, 8), because the hr size should be equivalent to HResult which is  a 32-bit value (Long).  A long is 4 bytes even on 64-bit operating systems.

The method StrCmpLogicalW is not present in Delphi compiler versions < 29.0. This issue has been added to MFPack WinApi.MediaFoundationApi.MfUtils.
So, unit uDirectoryTree uses clause needs to be expanded with Winapi.MediaFoundationApi.MfUtils;

By the way: I had to re assemble the whole project, because when compiling in DX7, no dcu's could be found in a path pointing to Madsi.

Regards Tony.

**Comment by FactoryXCode** [comment]
Ok, Got it working with Delphi XE2 up to XE7.

**Reply by rmesch** [reply]
Hi, Tony,
Sorry about the late reply, I've stopped all Delphi things for some time, but I'm back now.
I will incorporate your changes in the next version, I'm also updating it to the latest mfpack-version (not many changes needed).
The only Delphi-versions I have are Delphi 2006 and the latest community 12.1. The directory tree definitely works as is with Delphi 11 - 12.1, so I'm a bit puzzled.

I noticed that you have used my classes in the sinkwriter Example 3, with many interesting changes, some of which I will probably include in my next update, if you permit. I have some issues with it, though, which I'm reporting on the mfpack-rep.

Thanks,

Renate
